### PR TITLE
Add house routes

### LINKED ISSUE #14: /house/add
Inputs: name [string] + description [string]
Returns: `{"data": <house id>, "msg": <some message>, "status": <some status>}`

### LINKED ISSUE #15: /house/<id>/user/invite
Inputs: none
Returns: JSON object `{"data": <A generic invite link (anyone can join)>, "msg": <some message>, "status": <either "error" or "success" – no link is returned if the status is "error" – "data" will be blank>}`

### LINKED ISSUE #17: /house/<id>/user/join?token=<some token>
Inputs: none (note that this link is the one returned from the two endpoints above)
Returns: `{"data": <house id>, "msg": <some message>, "status": <"error" or "success">}`

### PR DESCRIPTION
**Pull request by teymour-aldridge**
Intends to fix #14, fix #15 and fix #17

### CODE REVIEW

File: app/house.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+house_blueprint = Blueprint("house", __name__, url_prefix="/house")
+
+
+@house_blueprint.route("/add", methods=("POST",))
+@jwt_required
+def add_house():
+    try:
+        name = request.json["name"]
+        description = request.json["description"]
```
Comment (gmelodie) [inline app/house.py:16]: Is `description` really mandatory?

File: app/house.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+@house_blueprint.route("/user/join")
+@jwt_required
+def join_house():
+    token = request.args["token"]
+    user = get_jwt_identity()
+    db_user: User = User.query.get(user)
+    if not user:
+        return (
+            jsonify(msg="The supplied token was invalid.", status="error", data=""),
```
Comment (gmelodie) [inline app/house.py:39]: Maybe return the house id in `data` (my idea of standarizing the responses)

File: app/house.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            400,
+        )
+    try:
+        try:
+            decoded_token = jwt.decode(token, current_app.config["SECRET_KEY"])
+        except jwt.DecodeError:
+            return (
+                jsonify(msg="The supplied token was invalid.", status="error", data=""),
+                400,
+            )
```
Comment (gmelodie) [inline app/house.py:49]: This doesn't need to be inside the nested `try`. Better:
```python
try:
    decoded_token = ...
except jwt.DecodeError:
    return ...

try: 
    if decoded_token["token_type"] == ...
...
```

File: app/house.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                )
+            else:
+                return jsonify(
+                    {
+                        "msg": "You cannot use this token – it is for someone else.",
+                        "status": "error",
+                        "data": "",
+                    }
+                )
+        elif decoded_token["token_type"] == "generic_join_house":
```
Comment (gmelodie) [inline app/house.py:69]: :scissors: `if` instead of `elif`

File: app/house.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                User.houses.append(House.query.get(decoded_token["house_id"]))
+                db.session.commit()
+                return jsonify(
+                    {
+                        "msg": "You are now in this house.",
+                        "data": decoded_token["house_id"],
+                        "status": "success",
+                    }
+                )
+            else:
```
Comment (gmelodie) [inline app/house.py:61]: :scissors: Don't need `else` as both clauses return

File: app/house.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            return jsonify(
+                {
+                    "msg": "You are now in this house.",
+                    "data": decoded_token["house_id"],
+                    "status": "success",
+                }
+            )
+    except jwt.DecodeError:
+        return (
+            jsonify(msg="The supplied token was invalid.", status="error", data=""),
```
Comment (gmelodie) [inline app/house.py:81]: Again, `data` to contain house id

File: app/house.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+def specific_email(house_id, identifier):
+    house = House.query.get(house_id)
+    if not house.users.filter(
+        (User.username == identifier) | (User.email == identifier)
+    ).first():
+        return jsonify(
+            {
+                "status": "error",
+                "msg": "You don't have permission to do this",
+                "data": "",
```
Comment (gmelodie) [inline app/house.py:96]: Again, `data` to contain house identifier `house_id`

File: app/house.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            }
+        )
+    user = User.filter(
+        (User.username == identifier) | (User.email == identifier)
+    ).first()
+    if not user:
+        return (
+            jsonify(
+                {"msg": "That user could not be found.", "status": "error", "data": ""}
```
Comment (gmelodie) [inline app/house.py:105]: Again, data: `house_id`

File: app/house.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            {
+                "msg": "Successfully created an invite link",
+                "status": "success",
+                "data": token,
+            }
+        )
+    else:
+        return (
+            jsonify(
+                status="error", msg="A house with that ID cannot be found", data=""
```
Comment (gmelodie) [inline app/house.py:128]: Data: ? :smile:

File: app/house.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+            token=jwt.encode(
+                {"token_type": "generic_join_house", "house_id": house.id},
+                current_app.config["SECRET_KEY"],
+            ),
+        )
+        return jsonify(data=token, msg="", status="success")
+    else:
+        return (
+            jsonify(
+                status="error", msg="A house with that ID cannot be found.", data=""
```
Comment (gmelodie) [inline app/house.py:164]: Data: `house_id`

File: app/house.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            current_app.config["SECRET_KEY"],
+        )
+        return jsonify(
+            {
+                "msg": "Successfully created an invite link",
+                "status": "success",
+                "data": token,
+            }
+        )
+    else:
```
Comment (gmelodie) [inline app/house.py:125]: :scissors: `else` (we never need `else` clauses to be honest)

File: app/house.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    user = get_jwt_identity()
+    db_user: User = User.query.get(user)
+    return_data = jsonify(
+        {
+            "data": list(map(lambda x: x.id, db_user.houses)),
+            "msg": "",
+            "status": "success",
+        }
+    )
+    return return_data
```
Comment (gmelodie) [inline app/house.py:146]: Include `msg` (`""`) and `status` (`success`)

File: app/house.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            "house.join_house",
+            token=jwt.encode(
+                {"token_type": "generic_join_house", "house_id": house.id},
+                current_app.config["SECRET_KEY"],
+            ),
+        )
+        return jsonify(data=token, msg="", status="success")
+    else:
+        return (
+            jsonify(
```
Comment (gmelodie) [inline app/house.py:163]: Here I think the inverted logic would be better:
```python
if not house:
    return #error
token = url_for ...
```

File: app/house.py (hunk 10 above)
Comment (teymour-aldridge) [reply app/house.py]: This is an error message –  so no data is returned.

File: app/house.py (hunk 9 above)
Comment (teymour-aldridge) [reply app/house.py:112]: It's intentionally left blank.

**Pull request merged**